Protect True and False Atom from .cleanup

brossignol opened the pull request:
When the `cleanup` method is used on any context the default Atom for True and False are cleaned too.

```
from pysat.formula import Formula, PYSAT_FALSE, PYSAT_TRUE
print(PYSAT_FALSE, PYSAT_TRUE)
>>>F T

Formula.cleanup()
print(PYSAT_FALSE, PYSAT_TRUE)
>>>None None
```

In particulars this mess up the results of simplified:
```
from pysat.formula import Atom, Or, Neg, Formula
print(Or(Atom(1), Neg(Atom(1))).simplified())
>>>T

Formula.cleanup()
print(Or(Atom(1), Neg(Atom(1))).simplified())
>>>None
```

I propose a fix to explicitly exclude PYSAT_FALSE and PYSAT_TRUE from cleanup and to register the same object again in the new context.

alexeyignatiev commented:
Well, I honestly am not sure I understand the need for the changes. When there is no context left after cleaning up, a new `'default'` context is created with the True and False constants reconstructed. Also, the note in [the documentation for `cleanup()`](https://pysathq.github.io/docs/html/api/formula.html#pysat.formula.Formula.cleanup) indicates that _"once cleaning of a context is done, the objects referring to the context’s formulas **must not be used**. At this point, they are orphaned and can’t get re-clausified"_.

brossignol commented:
`PYSAT_FALSE` and `PYSAT_TRUE` are **global** variable in formula.py and they are deleted by cleanup.
The variable `PYSAT_FALSE` and `PYSAT_TRUE` recreated in set_context line 589 are **local** variable, they do not recreate correctly the **global** variable `PYSAT_FALSE` and `PYSAT_TRUE`.

The problem is visible when simplified is used after cleanup, it returns the global variable.
```
Formula.cleanup()
print(Or(Atom(1), Neg(Atom(1))).simplified())
>>>None
```

Other solutions are possible:
* PYSAT_FALSE can be made into a local variable `self.PYSAT_FALSE`.
* PYSAT_FALSE can be replaced by a new instance when used `Atom(False)`.
* line 589 can set the global variable `global PYSAT_FALSE = Atom(False)`.

alexeyignatiev commented:
Right, I forgot to make them global... I feel that doing that may be the easiest and most sensible solution here. Let me think about the options you listed.

alexeyignatiev commented:
Okay, after thinking about this for a while, I believe the best solution would be to simply make the variables global when they are recreated in `set_context()`. Can you implement this option in your patch? I will merge it then.

brossignol commented:
I implemented the global variable solution.

It solves the result for `simplified`
```
Formula.cleanup()
print(Or(Atom(1), Neg(Atom(1))).simplified())
>>> T
```
But unfortunalty it doesn't solve the user import variables
```
from pysat.formula import Formula, PYSAT_FALSE, PYSAT_TRUE
Formula.cleanup()
print(PYSAT_FALSE, PYSAT_TRUE)
>>> None, None
```

The solution of protecting True/False from cleanup save the user imported variables.
There may be other ways, but I can't think of them.

alexeyignatiev commented:
I think the following can be done:
1. a context containing solely `PYSAT_FALSE` and `PYSAT_TRUE` can be created called, say, `'_global'` or `'_constants'`;
2. global constants are created once in the place they are currently created; this is done be temporarily setting the above context and then restoring the `'default'` context; this way the constants are stored in the above context specifically for them;
3.  method `Formula.__new__()` checks if the key is known first in the above context and then in the current context;
4. method 'set_context()` never updates these global constants;
5. method 'cleanup()` makes sure the above context is never cleaned up.
This solution looks fine to me except for step 3. I don't like that we have to make an additional check every time we create a new formula...

alexeyignatiev commented:
On the other hand, keeping a copy of these global constants in each context seems wasteful and adds up to a similar overhead when checking the existence of a newly obtained key. So these 5 steps may be the best solution for this problem.

brossignol commented:
Looks good to me.

Checking that a key is in a dict is an O(1) operation, I don't think step 3 add any measurable overhead.

alexeyignatiev commented:
OK, then let's proceed this way. I will try to find time for this at some point.

alexeyignatiev commented:
But apparently it fails the test you created for some reason...
